pennylane/devices/tests/test_compare_default_qubit.py

```diff
@@ -0,0 +1,211 @@
+
+        if dev.name == dev_def.name:
+            pytest.skip("Device is default.qubit.")
+
+        supports_tensor = (
+            "supports_tensor_observables" in dev.capabilities()
+            and dev.capabilities()["supports_tensor_observables"]
+        )
+
+        if not supports_tensor:
+            pytest.skip("Device does not support tensor observables.")
+
+        if not dev.analytic:
+            pytest.skip("Device is in non-analytical mode.")
+
+        n_layers = np.random.randint(1, 5)
+        weights = 2 * np.pi * np.random.rand(n_layers, 1)
+
+        def circuit(weights):
+            RandomLayers(weights, wires=range(n_wires))
+            return ret(qml.PauliZ(wires=0) @ qml.PauliX(wires=1))
+
+        qnode_def = qml.QNode(circuit, dev_def)
+        qnode = qml.QNode(circuit, dev)
+
+        grad_def = qml.grad(qnode_def, argnum=0)
+        grad = qml.grad(qnode, argnum=0)
+
+        assert np.allclose(qnode(weights), qnode_def(weights), atol=tol(dev.analytic))
+        assert np.allclose(grad(weights), grad_def(weights), atol=tol(dev.analytic))
+
+    def test_four_qubit_random_circuit(self, device, tol):
+        """Test a four-qubit random circuit with the whole set of possible gates"""
+        n_wires = 4
+        dev = device(n_wires)
+        dev_def = qml.device("default.qubit", wires=n_wires)
+
+        if dev.name == dev_def.name:
+            pytest.skip("Device is default.qubit.")
+
+        supports_tensor = (
+            "supports_tensor_observables" in dev.capabilities()
+            and dev.capabilities()["supports_tensor_observables"]
+        )
+
+        if not supports_tensor:
+            pytest.skip("Device does not support tensor observables.")
+
+        if not dev.analytic:
+            pytest.skip("Device is in non-analytical mode.")
+
+        gates = [
+            qml.PauliX,
+            qml.PauliY,
+            qml.PauliZ,
+            qml.S,
+            qml.T,
+            qml.RX,
+            qml.RY,
+            qml.RZ,
+            qml.Hadamard,
+            qml.Rot,
+            qml.CRot,
+            qml.Toffoli,
+            qml.SWAP,
+            qml.CSWAP,
+            qml.U1,
+            qml.U2,
+            qml.U3,
+            qml.CRX,
+            qml.CRY,
+            qml.CRZ,
+        ]
+
+        deep = 20
+        seed = random.randint(0, deep)
+
+        def circuit():
+            """4 qubits circuit with 20 random gates and random connections for
+            multi-qubit gates. The measurement is an arbitrary tensor product
+            over the qubits."""
+            random.seed(seed)
+            np.random.seed(seed)
+            for i in np.random.randint(0, len(gates), deep):
+                gate = gates[i]
+                params = list(np.pi * np.random.rand(gate.num_params))
+                gate(*params, wires=random.sample(range(n_wires), gate.num_wires))
+            return qml.expval(qml.PauliZ(0) @ qml.PauliX(1) @ qml.PauliY(2) @ qml.PauliZ(3))
```

Comment:
```suggestion
        layers = 3
        np.random.seed(1967)
        gates_per_layers = [np.random.permutation(gates) for _ in range(layers)]

        def circuit():
            """4-qubit circuit with layers of randomly selected gates and random connections for
            multi-qubit gates."""
            np.random.seed(1967)
            for gates in gates_per_layers:
                for gate in gates:
                    params = list(np.pi * np.random.rand(gate.num_params))
                    gate(
                        *params,
                        wires=np.random.choice(range(n_wires), size=gate.num_wires, replace=False)
                    )
            return qml.expval(qml.PauliZ(0))
```
I think this is a really nice test! Just proposing a tweak to the approach. The idea is that we have a couple of layers, and each layer is guaranteed to include every gate in `gates` (just a random permutation).